### ISSUE TITLE
performance improvements, where to start

### ISSUE THREAD

**Issue by bvdeenen** [Eonblast/Emysql#22]
Hi all

I'm running into some performance issues, and would like to ask for your input on where to start with trying to improve the emysql performance.

My table contains 5 small integer columns, but can have a million (or more rows). Doing "select \* from ..." with the normal mysql client takes about two seconds. Doing the same with emysql:execute takes about a 25 seconds. 

Do you already have any idea where the bottleneck is, or should I start profiling everything? Should I make my query return smaller chunks of data?

Just asking for some input, I do have the time to work on it.

Thanks 

Bart van Deenen, contractor at Spilgames.com

**Comment by hdiedrich** [comment]
Hi Bart, you are talking about a case when getting millions of result rows or any query?

Henning

**Comment by bvdeenen** [comment]
My immediate concern is the million records, and I'll probably work around that using mysql limit statements for getting chunks in parallel.
But looking through the code I found a place in the tcp module where I think there is some room for optimization. For all the queries I do, I translate the emysql resultset (basically a list of lists) into a list of records. It seems that if I could smuggle a custom function into the emysql_tcp module, I could skip a whole conversion step, although I have no idea how much time that would save. I'll definitely start doing some serious profiling. 
I'm also wondering if some of the lowest level functions could or should be implented as NIF's, but again, I'm going to measure first.

**Comment by bvdeenen** [comment]
I've worked around the millions of records stuff using a parallel job spawning mechanism, but I'm going to do some profiling on emysql soon, see if I can find some optimization targets. I'll keep you informed. We're going to be using emysql in a cluster with a combined system target of 1E6 queries per second, so performance means something :-)

**Comment by hdiedrich** [comment]
Hi Bart,

> I found a place in the tcp module where I think there is some room for optimization.

You are welcome to push your experiments as pull request, on a separate branch and I'll be happy to join the effort there.

Understand that in general, Erlang will not be as fast as C for wire protocol translation as it will usually copy more. If you do nothing else at all with the data then that is almost a contrived benchmark as that will be the part were Erlang will be slower than C (e.g. the mysql command line client). And you leave out the nice parts where the benefits come in.

Using NIFs sounds great! I would love to see a specialized driver like that. I believe that it should get its own branch to have the main driver stay pure Erlang. But to be able to switch to a speed optimized branch for special cases like your million rows would be exciting.

For the case you are describing, it is likely that the immutability of Erlang structures leads to a string of copying that is not helpful for performance when reading in millions of results. There are three cases here I would expect: most likely, there will be some optimizations that can be done, by avoiding type conversions and internal wrapping and copying of data; and which will be obvious when inspecting the code with this use case in mind, which is not what Emysql was originally build for. There will be things you can't get around except by NIFs. And there will be some things you can get around but might decide not to do so at least for Emysql proper, as they would go against the grain of Erlang. E.g. reducing everything to binaries, avoiding records, wrapping structures and function calls. You could do that but that will ruin the code and the gain in my view might not be worth that.

The question for all cases is if they are _really_ useful or only appear so in a test case that may not reflect actual use, as you may just have found with your process spawning solution. This sounds a lot like the Erlang way to me what you found there. What exactly did you do and what was the result?

However, I am all for performance, keep me posted and give me a chance to see your code.

Thanks,
Henning

**Comment by bvdeenen** [comment]
Ok, i'll be doing some profiling next week, I'm building cluster management tools right now. 

My quick fix is a module  that spawns a fixed number of workers in a queue, that each grab a subset of my million records. As soon as one is done, it talks back to the main process, and tells it to store the received rows in an ets table and spawns a new one. The first one that returns an empty reply terminates the whole pile of workers. The total throughput is about 3 times that of the direct "select \* from ..." emysql:execute call. But still, the 1E6 records is just a start. They are talking about 2E8 records with small integers, so it gets really interesting. Now in principle, I need just a subset of these per node, but the algorithm of which ones is hard to translate to MySQL (it uses the outcome of an erlang:phash2 together with a lookup table).

Looking through emysql_tcp I saw that the results are already in an ets table, so there's an easy performance benefit by skipping the ets->list->ets conversion. 

I'll be playing with this usecase extensively probably next week, and I'll keep you informed.

P.S. Is this issue tracker the best way to communicate on new development?

Thanks anyway you all that developed this library, it's good quality, and definitely something to build on.

**Comment by hdiedrich** [comment]
Since we don't have a mailing list, let's collect the notes here.

The hashing makes it sounds like you are testing a distributed cluster. I trust you are not emulating the cluster on one machine and try to reverse filter out the 'noise' to make it more real ;-)

If it is that easy to gain 3 times performance, it should be possible to detect the bottle neck and eliminate it.

**Comment by bvdeenen** [comment]
No. I'm testing on what we call the hairdryer, a pile of old Pentium-IV desktops (we call it the hairdryer because of the heat it generates, and the noise it starts making with the fans when I start loading it with tests :-). Fortunately my colleague Pedro who is in front of the fans is from Portugal, he can stand the heat :-) 

The second cluster is in a remote datacenter, and when in production it is supposed to handle a sustained throughput of about 1e6 queries per second to sharded databases. After that we'll go for multiple datacenters with partial datasets per datacenter. This company (Spilgames) is growing like a '90-ies startup.

**Comment by bvdeenen** [comment]
Hi all

I've done some performance measurements on emysql, comparing it with the mysql command line client as a baseline:

Conclusions:
- for tables with records with reasonable amounts of data per record (about 0.5kB per record) there is very little speed differences between emysql and mysql, only some 20% or so. This seems to indicate that it's mysql itself that is the bottleneck. I've tested this both on localhost as over a network.
- for tables with many small records with a few integers in them, the speedup is appreciable, mysql is about _7 times faster_ than emysql.
- scaling is linear, the time that emysql (or mysql) takes to process a certain number of rows is O(1) on the SQL limit parameter.

My conclusion is that I'm going to focus on the use-case with the small integers and very many rows. I've identified `emysql_tcp:decode_row_data/3` as a first target for optimization.
